### ISSUE TITLE
[Bug][AMD][diffusion] FLUX server crashes on ROCm: Triton PTX inline-asm ('=f' constraint) in diffusion layernorm/numerics kernels fails to compile on gfx942

### ISSUE THREAD

**Issue by kangwangamd** [sgl-project/sglang#34351]
## Summary
On ROCm (MI300X, gfx942), bringing up **FLUX.1-dev** with the native SGLang diffusion backend (`sglang serve --model-path black-forest-labs/FLUX.1-dev --model-type diffusion`) **crashes during warmup** with:

```
[aiter] import [module_rmsnorm_quant] under /sgl-workspace/aiter/aiter/jit/module_rmsnorm_quant.so
error: couldn't allocate output register for constraint 'f'
```

The server never becomes ready (`ready to roll` never prints, `/health_generate` stays `503`). In CI this manifests as `multimodal-gen-test-1-gpu-amd*` (`flux_image_t2i`) timing out at the 90-min step limit — see run [31327057626](https://github.com/sgl-project/sglang/actions/runs/31327057626) (PR Test ROCm 7.2 AMD, 08/09) and the 08/10 run 31415663809. It is deterministic and chronic.

## Root cause
The diffusion Triton fast-path helpers embed **NVIDIA PTX inline assembly** with float-register constraints (`"=f"`), which the AMDGPU backend cannot allocate (valid AMD constraints are `v`/`s`, not `f`):

- `python/sglang/kernels/ops/diffusion/triton/numerics.py`
  - L22-24 `rcp.approx.f32` … `constraints="=f,f,f"`
  - L35-37 `div.rn.f32 $0, $1, $2;` … `constraints="=f,f,f"`
  - L47-49 `rsqrt.approx.f32 $0, $1;` … `constraints="=f,f"` (+ `cuda_rsqrtf`)
- `python/sglang/kernels/ops/diffusion/triton/layernorm_modulate.py`
  - `_rcp4` (L~58-70): PTX block `rcp.approx.f32 / fma.rn.f32 / sub.ftz.f32` with `constraints="=f,f"`

These are pulled in via `fused_layernorm_modulate` on the FLUX DiT (`runtime/models/dits/flux.py`), JIT-compiled at the first warmup forward. There is **no ROCm/`is_hip` guard** on the dispatch, so ROCm takes the PTX path and the Triton compile aborts. (Note: the aiter CK `module_rmsnorm_quant` builds fine with `-DUSE_ROCM`; the "import module_rmsnorm_quant" line is just the last log line before the *Triton* compile crash — it is not the aiter kernel that fails.)

The `.cuh` C++ counterparts are correctly guarded (`dtype_float16.cuh` uses `#ifndef USE_ROCM … #else v_cvt_f16_f32 …`); these Triton `inline_asm_elementwise` helpers are the ones missing the guard.

## Repro (MI300X gfx942)
- Image: `rocm/sgl-dev:v0.5.17-rocm720-mi30x-20260809`; aiter commit `d9e5ef7ce`
- `sglang serve --model-path black-forest-labs/FLUX.1-dev --model-type diffusion --num-gpus 1`
- Crashes at warmup with the error above.
- **Gotcha:** passing a *local dir* path instead of the HF repo-id makes it fall back to the diffusers backend (which skips these kernels) and it appears to pass — the native path (repo-id) is required to hit the bug.

## Suggested fix
Provide a portable (non-PTX) implementation of these helpers and select it on ROCm — e.g. gate the `inline_asm_elementwise` PTX path behind `not is_hip()` and fall back to `tl.math` equivalents (`1.0/x`, `x/y`, `tl.rsqrt`) on AMD; or skip the fused-layernorm-modulate fast path on ROCm (the code already has an eager fallback via `torch.equal` verification for GLM-Image). Bit-exactness on AMD can use the native/eager path.

cc @BBuf (author of the diffusion layernorm/numerics fast paths, #3406 / Aug-08)

**Comment by sunxxuns** [comment]
Confirmed still present on current `main`, on a **public** image and **without
FLUX weights**. Three things to add: the files moved, there is a fifth affected
helper the report does not list, and the repro can be reduced to a ~40 s unit
test.

## Still failing at HEAD

Every PTX inline-asm helper on the diffusion fast path fails to compile on
gfx942 with the error in the report:

```
[ FAIL ] mul_rn_f32         common/numerics.py                  asm=mul.rn.f32             constraints==f,f,f
           error: couldn't allocate output register for constraint 'f'
[ FAIL ] div_rn_f32         common/numerics.py                  asm=div.rn.f32             constraints==f,f,f
           error: couldn't allocate output register for constraint 'f'
[ FAIL ] rsqrt_approx_f32   common/numerics.py                  asm=rsqrt.approx.f32       constraints==f,f
           error: couldn't allocate output register for constraint 'f'
[ FAIL ] cuda_rsqrtf        common/numerics.py                  wraps rsqrt.approx.f32     constraints==f,f
           error: couldn't allocate output register for constraint 'f'
[ FAIL ] _rcp4              norm/layernorm_modulate_triton.py   rcp.approx.f32 + fma.rn    constraints==f,f

=== 5 of 5 PTX helpers fail to compile on gfx942 ===
```

## Two corrections to the file list

**The paths changed.** #35114 ("Reorganize ops/diffusion by operator domain
behind a lazy facade", 2026-08-18) moved both files after this issue was filed.
The defect survived the move unchanged:

| in this report | at HEAD |
|---|---|
| `python/sglang/kernels/ops/diffusion/triton/numerics.py` | `python/sglang/kernels/ops/diffusion/common/numerics.py` |
| `python/sglang/kernels/ops/diffusion/triton/layernorm_modulate.py` | `python/sglang/kernels/ops/diffusion/norm/layernorm_modulate_triton.py` |

Neither file contains any `is_hip` / `USE_ROCM` guard at HEAD.

**`mul_rn_f32` is a fifth affected helper.** The report lists `rcp.approx.f32`,
`div.rn.f32`, `rsqrt.approx.f32` and `_rcp4`, but `common/numerics.py` also
exports `mul_rn_f32`, which emits `mul.rn.f32` with `constraints="=f,f,f"` and
fails identically. Whatever guard or lowering fixes the other four needs to
cover it too, otherwise the next caller to reach for a contraction-blocking
multiply hits the same wall.

## Repro without model weights

The original repro brings up a FLUX.1-dev server, but the failure is at Triton
JIT compile, so it reproduces on 256 elements with no checkpoint at all. Each
helper needs its own subprocess because the failure is a hard abort in the LLVM
AMDGPU backend, not a Python exception:

```python
import torch, triton
import triton.language as tl
from sglang.kernels.ops.diffusion.common.numerics import mul_rn_f32

@triton.jit
def k(x_ptr, y_ptr, o_ptr, n: tl.constexpr):
    i = tl.arange(0, n)
    tl.store(o_ptr + i, mul_rn_f32(tl.load(x_ptr + i), tl.load(y_ptr + i)))

N = 256
x = torch.rand(N, device="cuda") + 0.5
y = torch.rand(N, device="cuda") + 0.5
o = torch.empty(N, device="cuda")
k[(1,)](x, y, o, n=N)   # error: couldn't allocate output register for constraint 'f'
```

That makes a cheap CI guard possible on gfx942 — no 90-minute
`multimodal-gen-test-1-gpu-amd` job and no weights — which matters because the
timeout in that job is currently the only thing reporting this.

## Environment

| | |
|---|---|
| GPU | AMD Instinct MI300X, `gfx942:sramecc+:xnack-` |
| Image | `rocm/sgl-dev:v0.5.18-rocm720-mi30x-20260826` (public) |
| HIP | 7.2.26015-fc0010cf6a |
| Triton | 3.7.0 |
| PyTorch | 2.9.1+rocm7.2.0.git7e1940d4 |
| SGLang | `8eaffdf382` (2026-08-26) |

The report used `rocm/sgl-dev:v0.5.17-rocm720-mi30x-20260809`; this is the same
failure two SGLang minor versions later on a newer public image.
